File: src/tech/v3/dataset/rolling.clj
(ns tech.v3.dataset.rolling
  "Implement a generalized rolling window including support for time-based variable
  width windows."
  (:require [tech.v3.datatype :as dtype]
            [tech.v3.datatype.rolling :as dt-rolling]
            [tech.v3.datatype.argtypes :as argtypes]
            [tech.v3.datatype.datetime.operations :as dtype-dt-ops]
            [tech.v3.datatype.datetime :as dtype-dt]
            [tech.v3.datatype.statistics :as stats]
            [tech.v3.datatype.packing :as packing]
            [tech.v3.datatype.binary-op :as binary-op]
            [tech.v3.dataset.base :as ds-base])
  (:import [tech.v3.datatype Buffer])
  (:refer-clojure :exclude [min max nth first last]))


(defn mean
  [column-name]
  {:column-name column-name
   :reducer stats/mean
   :datatype :float64})


(defn sum
  [column-name]
  {:column-name column-name
   :reducer stats/sum
   :datatype :float64})


(defn min
  [column-name]
  {:column-name column-name
   :reducer stats/min})


(defn max
  [column-name]
  {:column-name column-name
   :reducer stats/max})


(defn variance
  [column-name]
  {:column-name column-name
   :reducer stats/variance
   :datatype :float64})


(defn standard-deviation
  [column-name]
  {:column-name column-name
   :reducer stats/standard-deviation
   :datatype :float64})


(defn nth
  "Get the nth window value"
  [column-name nth-val]
  {:column-name column-name
   :reducer (fn [rdr] (rdr nth-val))})


(defn first
  [column-name]
  {:column-name column-name
   :reducer (fn [rdr] (rdr 0))})


(defn last
  [column-name]
  {:column-name column-name
   :reducer (fn [rdr] (rdr -1))})


(defn ^:no-doc apply-window-ranges
  [ds windows reducer-map edge-mode]
  (->> reducer-map
       (map-indexed
        (fn [_idx [k red]]
          (assoc red :dest-column-name k)))
       (group-by :column-name)
       (mapv (fn [[colname reducers]]
               (let [colname (if (= :scalar (argtypes/arg-type colname))
                               [colname]
                               (vec colname))]
                 {:columns (mapv (partial ds-base/column ds) colname)
                  :reducers (vec reducers)})))
       (mapv
        (fn [{:keys [columns reducers]}]
          ;;common case is 1 column
          (if (== 1 (count columns))
            (let [win-data (dt-rolling/window-ranges->window-reader
                            (columns 0) windows edge-mode)]
              (mapv (fn [reducer]
                      {:tech.v3.dataset/name (:dest-column-name reducer)
                       :tech.v3.dataset/data
                       (-> (dtype/emap (:reducer reducer) (:datatype reducer :object)
                                       win-data)
                           (dtype/clone))})
                    reducers))
            (let [win-data (mapv #(dt-rolling/window-ranges->window-reader
                                   % windows edge-mode)
                                 columns)]
              (mapv (fn [reducer]
                      {:tech.v3.dataset/name (:dest-column-name reducer)
                       :tech.v3.dataset/data
                       (-> (apply dtype/emap (:reducer reducer) (:datatype reducer :object)
                                  win-data)
                           (dtype/clone))})
                    reducers)))))
       (apply concat)
       (reduce #(ds-base/add-column %1 %2) ds)))


(defn rolling
  "Perform a rolling window operation appending columns to the original dataset.

  * ds - src dataset.
  * window - either an integer for fixed window sizes or a map describing the window
    operation containing keys:
    - `:window-type` - either `:fixed` or `:variable`.  For variable window operations
      `:column-name` must be a monotonically increasing column.
    - `:window-size` - for fixed window operation must be a positive integer.  For
       variable window operations must be a double value which is produced via a
       comparison function.
    - `:relative-window-position` - describes where the window is
       positioned.  Operations are `:left`, `:center`, `:right` and defaults to
       `:center` for fixed and `:right` for relative window types.
    - `:edge-mode` - for fixed windows describes what values to fill in at the edges
       of the source column.  Options are `:zero` which is 0 for numeric types and `nil`
       for object types and `:clamp` which fills in the first,last values of the column
       respectively.  Defaults to `:clamp`.
    - `:comp-fn` - if provided must return a double which is the result of comparing
      the last value of the range to the first which means `clojure.core/-`
      is a reasonable default.
    - `:units` - for datetime types, describes the units of `:window-size` and will
      dictate the numeric space if `:comp-fn` is not provided.
  * reducer-map - A map of result column name to reducer map.  The reducer map is a
    map which must contain at least `{:column-name :reducer}` where reducer is an ifn
    that is passed each window.  The result column is scanned to ascertain datatype and
    missing value status.  Multi-column reducers are supported if column-name is a vector
    of column names.  In that case each column's window is passed to the reducer.  The
    reducer can also specify the final datatype if `:datatype` is a key in the map.  Beware,
    however, that this disables missing value detection for integer datatypes.


**Fixed Window Examples:**


```clojure
user> (def test-ds (ds/->dataset {:a (map #(Math/sin (double %))
                                          (range 0 200 0.1))}))
#'user/test-ds
user> (ds/head (ds-roll/rolling test-ds 10 {:mean (ds-roll/mean :a)
                                            :min (ds-roll/min :a)
                                            :max (ds-roll/max :a)}))
_unnamed [5 4]:

|         :a |      :mean | :min |       :max |
|-----------:|-----------:|-----:|-----------:|
| 0.00000000 | 0.09834413 |  0.0 | 0.38941834 |
| 0.09983342 | 0.14628668 |  0.0 | 0.47942554 |
| 0.19866933 | 0.20275093 |  0.0 | 0.56464247 |
| 0.29552021 | 0.26717270 |  0.0 | 0.64421769 |
| 0.38941834 | 0.33890831 |  0.0 | 0.71735609 |
user> (ds/head (ds-roll/rolling test-ds
                                {:window-type :fixed
                                 :window-size 10
                                 :relative-window-position :left}
                                {:mean (ds-roll/mean :a)
                                 :min (ds-roll/min :a)
                                 :max (ds-roll/max :a)}))
_unnamed [5 4]:

|         :a |      :mean | :min |       :max |
|-----------:|-----------:|-----:|-----------:|
| 0.00000000 | 0.00000000 |  0.0 | 0.00000000 |
| 0.09983342 | 0.00998334 |  0.0 | 0.09983342 |
| 0.19866933 | 0.02985027 |  0.0 | 0.19866933 |
| 0.29552021 | 0.05940230 |  0.0 | 0.29552021 |
| 0.38941834 | 0.09834413 |  0.0 | 0.38941834 |
user> (ds/head (ds-roll/rolling test-ds
                                {:window-type :fixed
                                 :window-size 10
                                 :relative-window-position :right}
                                {:mean (ds-roll/mean :a)
                                 :min (ds-roll/min :a)
                                 :max (ds-roll/max :a)}))
_unnamed [5 4]:

|         :a |      :mean |       :min |       :max |
|-----------:|-----------:|-----------:|-----------:|
| 0.00000000 | 0.41724100 | 0.00000000 | 0.78332691 |
| 0.09983342 | 0.50138810 | 0.09983342 | 0.84147098 |
| 0.19866933 | 0.58052549 | 0.19866933 | 0.89120736 |
| 0.29552021 | 0.65386247 | 0.29552021 | 0.93203909 |
 | 0.38941834 | 0.72066627 | 0.38941834 | 0.96355819 |

user> ;;Multi column reducer
user> (ds/head (ds-roll/rolling test-ds 10
                                {:c {:column-name [:a :a]
                                     :reducer (fn [a b]
                                                (Math/round
                                                 (+ (dfn/sum a) (dfn/sum b))))
                                     :datatype :int16}}))
_unnamed [5 2]:

|         :a | :c |
|-----------:|---:|
| 0.00000000 |  2 |
| 0.09983342 |  3 |
| 0.19866933 |  4 |
| 0.29552021 |  5 |
| 0.38941834 |  7 |
```

**Variable Window Examples:**


```clojure
user> (def stocks (ds/->dataset \"test/data/stocks.csv\" {:key-fn keyword}))
#'user/stocks
user> ;;variable window column must be monotonically increasing
user> (def stocks (ds/sort-by-column stocks :date))
#'user/stocks
user> (ds/head stocks)
test/data/stocks.csv [5 3]:

| :symbol |      :date | :price |
|---------|------------|-------:|
|    AAPL | 2000-01-01 |  25.94 |
|     IBM | 2000-01-01 | 100.52 |
|    MSFT | 2000-01-01 |  39.81 |
|    AMZN | 2000-01-01 |  64.56 |
|    AAPL | 2000-02-01 |  28.66 |

user> (ds/head (ds-roll/rolling stocks
                                {:window-type :variable
                                 :column-name :date
                                 :units :days
                                 :window-size 3}
                                {:price-mean-3d (ds-roll/mean :price)
                                 :price-max-3d (ds-roll/max :price)
                                 :price-min-3d (ds-roll/min :price)}))
test/data/stocks.csv [5 6]:

| :symbol |      :date | :price | :price-mean-3d | :price-max-3d | :price-min-3d |
|---------|------------|-------:|---------------:|--------------:|--------------:|
|    AAPL | 2000-01-01 |  25.94 |    57.70750000 |        100.52 |         25.94 |
|     IBM | 2000-01-01 | 100.52 |    68.29666667 |        100.52 |         39.81 |
|    MSFT | 2000-01-01 |  39.81 |    52.18500000 |         64.56 |         39.81 |
|    AMZN | 2000-01-01 |  64.56 |    64.56000000 |         64.56 |         64.56 |
|    AAPL | 2000-02-01 |  28.66 |    56.49750000 |         92.11 |         28.66 |
user> (ds/head (ds-roll/rolling stocks
                                {:window-type :variable
                                 :column-name :date
                                 :units :months
                                 :window-size 3}
                                {:price-mean-3d (ds-roll/mean :price)
                                 :price-max-3d (ds-roll/max :price)
                                 :price-min-3d (ds-roll/min :price)}))
test/data/stocks.csv [5 6]:

| :symbol |      :date | :price | :price-mean-3d | :price-max-3d | :price-min-3d |
|---------|------------|-------:|---------------:|--------------:|--------------:|
|    AAPL | 2000-01-01 |  25.94 |    58.92500000 |        106.11 |         25.94 |
|     IBM | 2000-01-01 | 100.52 |    61.92363636 |        106.11 |         28.66 |
|    MSFT | 2000-01-01 |  39.81 |    58.06400000 |        106.11 |         28.66 |
|    AMZN | 2000-01-01 |  64.56 |    60.09222222 |        106.11 |         28.66 |
|    AAPL | 2000-02-01 |  28.66 |    57.56583333 |        106.11 |         28.37 |
```"
  ([ds window reducer-map _options]
   (let [n-rows (ds-base/row-count ds)
         window-data (if (integer? window)
                       {:window-size window
                        :relative-position :center
                        :window-type :fixed}
                       window)
         windows
         (case (:window-type window-data :fixed)
           :fixed
           (dt-rolling/fixed-rolling-window-ranges
            n-rows (:window-size window-data)
            (:relative-window-position window-data :center))
           :variable
           (let [_ (when-not (:column-name window-data)
                     (throw (Exception. (format "Variable rolling windows must have :column-name in the window data"))))
                 src-col (ds-base/column ds (:column-name window-data))
                 col-dt (dtype/elemwise-datatype src-col)]
             (vec (dt-rolling/variable-rolling-window-ranges
                   src-col (:window-size window-data)
                   {:comp-fn
                    (if-let [comp-fn (:comp-fn window-data)]
                      comp-fn
                      (when (dtype-dt/datetime-datatype? (packing/unpack-datatype col-dt))
                        (dtype-dt-ops/between-op
                         (dtype/elemwise-datatype src-col)
                         (:units window-data :milliseconds)
                         true)))
                    :relative-window-position (get window-data :relative-window-position)}))))]
     (apply-window-ranges ds windows reducer-map (:edge-mode window-data :clamp))))
  ([ds window reducer-map]
   (rolling ds window reducer-map nil)))


(defn expanding
  "Run a set of reducers across a dataset with an expanding set of windows.  These
  will produce a cumsum-type operation."
  [ds reducer-map]
  (apply-window-ranges ds (dt-rolling/expanding-window-ranges
                           (ds-base/row-count ds))
                       reducer-map
                       :clamp))

(comment
  (require '[tech.v3.dataset :as ds])
  (def test-ds (ds/->dataset {:a (map #(Math/sin (double %))
                                      (range 0 200 0.1))}))
  (rolling test-ds 10 {:mean (mean :a)
                       :min (min :a)
                       :max (max :a)})
  )

File: java_public_api/tech/v3/dataset/Rolling.java
package tech.v3.dataset;

import static tech.v3.Clj.*;
import clojure.lang.IFn;
import clojure.lang.Keyword;
import java.util.Map;

/**
 * Fixed and variable length rolling windows.  For variable rolling windows the dataset
 * must already be sorted by the target column.  Datetime support is provided in terms of
 * provide specific units in which to perform the rolling operation such as the keyword
 * `:days`.
 *
 */
public class Rolling {

  private Rolling(){}

  static final IFn meanFn = requiringResolve("tech.v3.dataset.rolling", "mean");
  static final IFn sumFn = requiringResolve("tech.v3.dataset.rolling", "sum");
  static final IFn minFn = requiringResolve("tech.v3.dataset.rolling", "min");
  static final IFn maxFn = requiringResolve("tech.v3.dataset.rolling", "max");
  static final IFn varianceFn = requiringResolve("tech.v3.dataset.rolling", "variance");
  static final IFn stddevFn = requiringResolve("tech.v3.dataset.rolling", "standard-deviation");
  static final IFn nth = requiringResolve("tech.v3.dataset.rolling", "nth");
  static final IFn firstFn = requiringResolve("tech.v3.dataset.rolling", "first");
  static final IFn lastFn = requiringResolve("tech.v3.dataset.rolling", "last");
  static final IFn rollingFn = requiringResolve("tech.v3.dataset.rolling", "rolling");

  /**
   * Fixed or variable rolling window reductions.
   *
   * @param windowSpec Window specification specifying the type of window, either a
   * window over a fixed number of rows or a window based on a fixed logical
   * quantitative difference i.e. three months or 10 milliseconds.
   * @param reducerMap map of dest column name to reducer where reducer is a map with
   * two keys, :column-name which is the input column to use and :reducer which is
   * an IFn that receives each window of data as a buffer.
   *
   * Example:
   *
   *```java
   * Map stocks = makeDataset("https://github.com/techascent/tech.ml.dataset/raw/master/test/data/stocks.csv");
   *
   * //Variable-sized windows require the source column to be sorted.
   * stocks = sortByColumn(stocks, "date");
   * Map variableWin = Rolling.rolling(stocks,
   *				      Rolling.variableWindow("date", 3, kw("months")),
   *				      hashmap("price-mean-3m", Rolling.mean("price"),
   *					      "price-max-3m", Rolling.max("price"),
   *					      "price-min-3m", Rolling.min("price")));
   *println(head(variableWin, 10));
   *https://github.com/techascent/tech.ml.dataset/raw/master/test/data/stocks.csv [10 6]:
   * //| symbol |       date |  price | price-max-3m | price-mean-3m | price-min-3m |
   * //|--------|------------|-------:|-------------:|--------------:|-------------:|
   * //|   AAPL | 2000-01-01 |  25.94 |       106.11 |   58.92500000 |        25.94 |
   * //|    IBM | 2000-01-01 | 100.52 |       106.11 |   61.92363636 |        28.66 |
   * //|   MSFT | 2000-01-01 |  39.81 |       106.11 |   58.06400000 |        28.66 |
   * //|   AMZN | 2000-01-01 |  64.56 |       106.11 |   60.09222222 |        28.66 |
   * //|   AAPL | 2000-02-01 |  28.66 |       106.11 |   57.56583333 |        28.37 |
   * //|   MSFT | 2000-02-01 |  36.35 |       106.11 |   60.19363636 |        28.37 |
   * //|    IBM | 2000-02-01 |  92.11 |       106.11 |   62.57800000 |        28.37 |
   * //|   AMZN | 2000-02-01 |  68.87 |       106.11 |   59.29666667 |        28.37 |
   * //|   AMZN | 2000-03-01 |  67.00 |       106.11 |   54.65583333 |        21.00 |
   * //|   MSFT | 2000-03-01 |  43.22 |       106.11 |   53.53363636 |        21.00 |
   *
   * //Fixed window...
   *
   * Object radians = VecMath.mul(2.0*Math.PI, VecMath.div(range(33), 32.0));
   * Map sinds = makeDataset(hashmap("radians", radians, "sin", VecMath.sin(radians)));
   * Map fixedWin = Rolling.rolling(sinds,
   *				   Rolling.fixedWindow(4),
   *				   hashmap("sin-roll-mean", Rolling.mean("sin"),
   *					   "sin-roll-max", Rolling.max("sin"),
   *					   "sin-roll-min", Rolling.min("sin")));
   *println(head(fixedWin, 8));
   * //_unnamed [8 5]:

   * //|        sin |    radians | sin-roll-max | sin-roll-min | sin-roll-mean |
   * //|-----------:|-----------:|-------------:|-------------:|--------------:|
   * //| 0.00000000 | 0.00000000 |   0.19509032 |   0.00000000 |    0.04877258 |
   * //| 0.19509032 | 0.19634954 |   0.38268343 |   0.00000000 |    0.14444344 |
   * //| 0.38268343 | 0.39269908 |   0.55557023 |   0.00000000 |    0.28333600 |
   * //| 0.55557023 | 0.58904862 |   0.70710678 |   0.19509032 |    0.46011269 |
   * //| 0.70710678 | 0.78539816 |   0.83146961 |   0.38268343 |    0.61920751 |
   * //| 0.83146961 | 0.98174770 |   0.92387953 |   0.55557023 |    0.75450654 |
   * //| 0.92387953 | 1.17809725 |   0.98078528 |   0.70710678 |    0.86081030 |
   * //| 0.98078528 | 1.37444679 |   1.00000000 |   0.83146961 |    0.93403361 |
   *```
   */
  public static Map rolling(Object ds, Map windowSpec, Map reducerMap) {
    return (Map)rollingFn.invoke(ds, windowSpec, reducerMap);
  }
  /**
   * Create a variable window specification with a double windowsize for a particular column.
   * This specification will not work on datetime columns.
   */
  public static Map variableWindow(Object colname, double windowSize) {
    return hashmap(kw("window-type"), kw("variable"),
		   kw("column-name"), colname,
		   kw("window-size"), windowSize);
  }
  /**
   * Create a variable window specification with a double windowsize for a particular column
   * and a compFn which must take two values and return a double.  The function must take 2
   * arguments and the arguments are passed in as (later,earlier).  This allows the basic
   * clojure '-' operator to work fine in many cases.
   *
   */
  public static Map variableWindow(Object colname, double windowSize, Object compFn) {
    return hashmap(kw("window-type"), kw("variable"),
		   kw("column-name"), colname,
		   kw("window-size"), windowSize,
		   kw("comp-fn"), compFn);
  }
  /**
   * Create a datetime-specific variable window specification with a double windowsize for
   * a particular column.
   *
   * @param datetimeUnit One of `[:milliseconds, :seconds, :hours, :days, :months]`.
   */
  public static Map variableWindow(Object colname, double windowSize, Keyword datetimeUnit) {
    return hashmap(kw("window-type"), kw("variable"),
		   kw("column-name"), colname,
		   kw("window-size"), windowSize,
		   kw("units"), datetimeUnit);
  }
  /**
   * Return fixed size rolling window.  Window will be fixed over `window-size` rows.
   */
  public static Map fixedWindow(long windowSize) {
    return hashmap(kw("window-type"), kw("fixed"),
		   kw("window-size"), windowSize);
  }
  /**
   * Return fixed size rolling window.  Window will be fixed over `window-size` rows.
   *
   * @param winPos One of `[:left :center :right]`.  This combined with the default
   *        edge mode of `:clamp` dictates the windows of data the reducer sees.
   */
  public static Map fixedWindow(long windowSize, Keyword winPos) {
    return hashmap(kw("window-type"), kw("fixed"),
		   kw("window-size"), windowSize,
		   kw("relative-window-position"), winPos);
  }
  /**
   * Return fixed size rolling window.  Window will be fixed over `window-size` rows.
   *
   * @param winPos One of `[:left :center :right]`.  This combined with the default
   *        edge mode dictates windows of data the reducer sees.
   *
   * @param edgeMode One of `[:zero, null, :clamp]`.  Clamp means repeat the end value.
   */
  public static Map fixedWindow(long windowSize, Keyword winPos, Keyword edgeMode) {
    return hashmap(kw("window-type"), kw("fixed"),
		   kw("window-size"), windowSize,
		   kw("relative-window-position"), winPos,
		   kw("edge-mode"), edgeMode);
  }
  /**
   * Create a columnwise reducer.  This reducer gets sub-windows from the column and
   * must return a scalar value.
   */
  public static Map reducer(Object srcColname, IFn reduceFn) {
    return hashmap(kw("column-name"), srcColname,
		   kw("reducer"), reduceFn);
  }
  /** mean reducer*/
  public static Map mean(Object colname) {
    return (Map)meanFn.invoke(colname);
  }
  /** sum reducer*/
  public static Map sum(Object colname) {
    return (Map)sumFn.invoke(colname);
  }
  /** min reducer*/
  public static Map min(Object colname) {
    return (Map)minFn.invoke(colname);
  }
  /** max reducer*/
  public static Map max(Object colname) {
    return (Map)maxFn.invoke(colname);
  }
  /** stddev reducer*/
  public static Map stddev(Object colname) {
    return (Map)stddevFn.invoke(colname);
  }
  /** variance reducer*/
  public static Map variance(Object colname) {
    return (Map)varianceFn.invoke(colname);
  }
  /** reducer that keeps the first value*/
  public static Map first(Object colname) {
    return (Map)firstFn.invoke(colname);
  }
  /** reducer that keeps the last value*/
  public static Map last(Object colname) {
    return (Map)lastFn.invoke(colname);
  }
}
